### PR TITLE
feat: implement duplicate parameter type check

### PR DESCRIPTION
**Pull request by DaniPopes**
part 1 of 100000 of type checking

### CODE REVIEW

File: crates/sema/src/ty/mod.rs [hunk 1]
```diff
@@ -238,28 +248,44 @@ impl<'gcx> Gcx<'gcx> {
...
-            _ => return None,
-        })
+        let params = self.item_parameters(id);
+        debug_assert_eq!(tys.len(), params.len());
+        std::iter::zip(tys.iter().copied(), params.iter().copied())
+    }
+
+    /// Returns the parameter variable declarations of the given function-like item.
+    ///
+    /// Also accepts structs.
```
Comment (onbjerg) [inline crates/sema/src/ty/mod.rs:263]: nit: document panic

File: crates/sema/src/ty/mod.rs [hunk 2]
```diff
@@ -238,28 +248,44 @@ impl<'gcx> Gcx<'gcx> {
...
+    pub fn item_parameters_opt(self, id: hir::ItemId) -> Option<&'gcx [hir::VariableId]> {
+        self.hir.item(id).parameters()
+    }
+
+    /// Returns the return variable declarations of the given function-like item.
```
Comment (onbjerg) [inline crates/sema/src/ty/mod.rs:280]: nit: document panic

File: crates/sema/src/typeck/mod.rs [hunk 3]
```diff
@@ -1,22 +1,83 @@
...
+                return false;
+            }
+        }
+        if !same_external_params(gcx, gcx.type_of_item(a), gcx.type_of_item(b)) {
+            return false;
+        }
+        true
+    };
+
+    reported.clear();
```
Comment (onbjerg) [inline crates/sema/src/typeck/mod.rs]: no need to clear here its already empty

File: tests/ui/typeck/duplicate_overloaded_items.sol [hunk 4]
```diff
@@ -0,0 +1,49 @@
+contract C {
```
Comment (onbjerg) [inline tests/ui/typeck/duplicate_overloaded_items.sol:50]: we should add tests for top-level events and fns with duplicate names

File: crates/sema/src/typeck/mod.rs [hunk 5]
```diff
@@ -1,22 +1,83 @@
...
+    for (_name, decls) in &scope.declarations {
+        let decls = &decls[..];
+        if decls.len() <= 1 {
+            continue;
+        }
+        for (i, &decl) in decls.iter().enumerate() {
```
Comment (onbjerg) [inline crates/sema/src/typeck/mod.rs:52]: we should clear in here before this loop though
